### DISCUSSION TITLE
Profiling Mitsuba on Linux with AMD CPU

### DISCUSSION THREAD

**Discussion by frettini** [mitsuba-renderer/mitsuba3#1200] [Q&A] [answered]
Hi all!
I am implementing a custom integrator and I would like to profile my implementation to identify bottlenecks. I saw that Mitsuba comes with a profiler that either uses IttNotify and VTune or NvTools and Nsight. I am focusing on CPU applications, but I am unsure I can use VTune as I have an AMD Ryzen 9 5900X (VTune is unable to find my processor, I imagine because it is not Intel).  Could you point me to a way to profile my implementation? This would be for either scalar or llvm variants. Many thanks.

**Comment by njroussel** [comment, marked as answer]
Hi @frettini 

For scalar code, I've had success using `perf` and whatever GUI you prefer ([Firefox Profiler](https://profiler.firefox.com/) is my personal favorite).
For LLVM, it's a bit more tricky. We don't instrument the LLVM kernels at all AFAIK, so whatever hotspots you could find would be in the x86 binary and you'd need to reverse-engineer/guess what the original source code is. Personally, I've only ever done this for microbenchmarks, where the assembled x86 was fairly small. I don't think this is suitable for profiling an entire integrator.

**Reply by frettini** [reply]
Hello,
I've just tried out perf and it works quite well for my use. Many Thanks!
